src/axom/quest/SamplingShaper.hpp

```diff
@@ -724,13 +553,263 @@ class SamplingShaper : public Shaper
+        }
+        else if(meshDim == 3)
+        {
+          shaper->template sampleInOutField<3>(m_dc,
+                                               m_inoutShapeQFuncs,
+                                               m_quadratureOrder,
+                                               m_projector33);
+        }
+        break;
+      }
+      break;
+    case shaping::VolFracSampling::SAMPLE_AT_DOFS:
+      SLIC_ERROR("Not implemented yet!");
+      break;
+    }
+  }
+
+  /**
+   * \brief Compute volume fractions for a given material using its associated quadrature function.
+   * 
+   * The generated grid function will be registered in the data collection and prefixed by `vol_frac_`
+   *
+   * \param [in] matField The name of the material
+   */
+  void computeVolumeFractionsForMaterial(const std::string& matField)
+  {
+    AXOM_ANNOTATE_SCOPE("computeVolumeFractionsForMaterial");
+
+    // Retrieve the inout samples QFunc
+    SLIC_ASSERT(axom::utilities::string::startsWith(matField, "mat_inout_"));
+    mfem::QuadratureFunction* inout = m_inoutMaterialQFuncs.Get(matField);
+
+    const auto& ir = inout->GetSpace()->GetIntRule(0);  // assume all elements are the same
+    const int sampleOrder = ir.GetOrder();
+    const int sampleNQ = ir.GetNPoints();
+    const int sampleSZ = inout->GetSpace()->GetSize();
+
+    // print info about sampling on rank 0
+    // TODO: mpi reduce this for stats on all ranks
+    SLIC_INFO_ROOT(axom::fmt::format(axom::utilities::locale(),
+                                     "In computeVolumeFractions(): sample order {} | "
+                                     "sample num qpts {} |  total samples {:L}",
+                                     sampleOrder,
+                                     sampleNQ,
+                                     sampleSZ));
+
+    // extract some properties from computational mesh
+    mfem::Mesh* mesh = m_dc->GetMesh();
+    const int dim = mesh->Dimension();
+    const int NE = mesh->GetNE();
+    SLIC_INFO_ROOT(
+      axom::fmt::format(axom::utilities::locale(), "Mesh has dim {} and {:L} elements", dim, NE));
+
+    // Access or create a registered volume fraction grid function from the data collection
+    const auto vf_name = axom::fmt::format("vol_frac_{}", matField.substr(10));
+    mfem::GridFunction* vf = shaping::getOrAllocateL2GridFunction(m_dc,
+                                                                  vf_name,
+                                                                  m_volfracOrder,
+                                                                  dim,
+                                                                  mfem::BasisType::Positive);
+    const mfem::FiniteElementSpace* fes = vf->FESpace();
+    const int dofs = fes->GetFE(0)->GetDof();
+
+    // access or compute the mass matrix
+    mfem::DenseTensor* mass_mat {nullptr};
+    const std::string mass_matrix_name = "shaping_mass_matrix";
```

Comment:
Do these names ever change? Should they be class members with these defaults, in other words?

Reply:
I don't anticipate the names changing. 
These objects are created and used only in this function, but the function is called once for each material.